cirq-google/cirq_google/transformers/target_gatesets/sycamore_gateset.py

```diff
@@ -0,0 +1,153 @@
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Target gateset used for compiling circuits to Sycamore + 1-q rotations + measurement gates."""
+import itertools
+from typing import List, Optional, Sequence
+
+import cirq
+from cirq.protocols.decompose_protocol import DecomposeResult
+from cirq.transformers.target_gatesets.compilation_target_gateset import (
+    _create_transformer_with_kwargs,
+)
+from cirq_google import ops
+from cirq_google.transformers.analytical_decompositions import two_qubit_to_sycamore
+
+
+@cirq.transformer
+def merge_swap_rzz_and_2q_unitaries(
+    circuit: 'cirq.AbstractCircuit',
+    *,
+    context: Optional['cirq.TransformerContext'] = None,
+    merged_swap_rzz_tag: str = "_merged_swap_rzz",
+    merged_2q_component_tag: str = "_merged_2q_unitaries",
+) -> 'cirq.Circuit':
+    """Merges 2-qubit connected components and adjacent `cirq.SWAP` and `cirq.ZZPowGate` gates.
+
+    Does the following two transformations, in that order:
+    1. Merges adjacent occurrences of `cirq.SWAP` and `cirq.ZZPowGate` into a
+    `cirq.CircuitOperation` tagged with `merged_swap_rzz_tag`.
+    2. Merges connected components of 1 and 2 qubit unitaries in the circuit into a
+    `cirq.CircuitOperation` tagged with `merged_2q_component_tag`, ignoring the newly
+    introduced tagged circuit operations added in Step-1.
+
+    Args:
+        circuit: Input circuit to transform. It will not be modified.
+        context: `cirq.TransformerContext` storing common configurable options for transformers.
+        merged_swap_rzz_tag: Tag to apply on newly introduced circuit operations wrapping adjacent
+            `cirq.SWAP` and `cirq.ZZPowGate`s.
+        merged_2q_component_tag: Tag to apply on newly introduced circuit operations wrapping
+            connected components of 1 and 2 qubit unitaries.
+
+    Returns:
+        Copy of the transformed input circuit.
+
+    Raises:
+          ValueError: If merged_2q_component_tag and merged_swap_rzz_tag are equal.
+    """
+    if merged_2q_component_tag == merged_swap_rzz_tag:
+        raise ValueError("merged_swap_rzz_tag and merged_2q_component_tag should be different.")
+
+    def merge_func_swap_rzz(
+        ops1: Sequence['cirq.Operation'], ops2: Sequence['cirq.Operation']
+    ) -> bool:
+        if not (len(ops1) == 1 and len(ops2) == 1):
+            return False
+        for op1, op2 in itertools.permutations([ops1[0], ops2[0]]):
+            if op1.gate == cirq.SWAP and isinstance(op2.gate, cirq.ZZPowGate):
+                return True
+        return False
+
+    tags_to_ignore = context.tags_to_ignore if context else ()
+    circuit = cirq.merge_operations_to_circuit_op(
+        circuit,
+        merge_func_swap_rzz,
+        tags_to_ignore=tags_to_ignore,
+        merged_circuit_op_tag=merged_swap_rzz_tag,
```

Comment:
Why do we tag these connected components? Is it for users to read the circuit and identify transformation steps?

Reply:
Yes, the tags are useful in logging but also useful for subsequent transformation steps where we want to identify whether a newly created connected component is a swap + rzz component (and hence is a known decomposition) or is an arbitrary 2q connected component and should be decomposed analytically or using the tabulation.